File: base/sort/python/BubbleSort.py
# 时间复杂度O(n^2),最坏情况O(n^2)，最好情况O(n),空间复杂度O(1),稳定
def bubble_sort(unsort_nums):
    for i in range(len(unsort_nums) - 1):
        for j in range(len(unsort_nums) - i - 1):
            print(i, j+1)
            if unsort_nums[j] > unsort_nums[j + 1]:
                unsort_nums[j], unsort_nums[j + 1] = unsort_nums[j + 1], unsort_nums[j]

File: base/sort/java/BubbleSort.java
package sort.java;

//冒泡排序
public class BubbleSort {

    public static void bubbleSort(int[] array) {
        boolean didSwap = false;
        for (int i = 0; i < array.length - 1; i++) {
            //每轮比较完，最大的数沉到底部
            for (int j = 0; j < array.length - i - 1; j++) {
                if (array[j] > array[j + 1]) {
                    swap(array, j, j + 1);
                    didSwap = true;
                }
            }
            if(!didSwap) {
                return;
            }
        }
    }

    private static void swap(int[] array, int i, int j) {
        int tmp = array[i];
        array[i] = array[j];
        array[j] = tmp;
    }
}
